### ISSUE TITLE
Duplicated OTel spans on the new grpc server implementation.

### ISSUE THREAD

**Issue by brunobat** [quarkusio/quarkus#37140]
### Describe the bug

This issue was first reported by @edeandrea on this zulip thread: https://quarkusio.zulipchat.com/#narrow/stream/187030-users/topic/gRPC.20server.20oTel.20errors/near/398274468

On the initial report there was scope.close warnings on the logs:
```log
07:57:59 WARN  [io.qu.op.ru.QuarkusContextStorage] (vert.x-eventloop-thread-3) Context in storage not the expected context, Scope.close was not called correctly. Details: OTel context before: {spanId=6ab0c0488d656975, traceId=11cf803891be9354f0e1e5e7b434cda6, sampled=true, parentId=b59e2af98fc4046d}. OTel context toAttach: {spanId=b59e2af98fc4046d, traceId=11cf803891be9354f0e1e5e7b434cda6, sampled=true}
07:57:59 WARN  [io.qu.op.ru.QuarkusContextStorage] (vert.x-eventloop-thread-3) Context in storage not the expected context, Scope.close was not called correctly. Details: OTel context before: {}. OTel context toAttach: {spanId=6ab0c0488d656975, traceId=11cf803891be9354f0e1e5e7b434cda6, sampled=true, parentId=b59e2af98fc4046d}
```

Subsequent analysis revealed that the spans for the grpc server were being duplicated because 2 different OTel instrumenters were in the request pipeline:
* HttpInstrumenterVertxTracer
* GrpcTracingServerInterceptor
Each instrumenter will create a span and sometimes they end the spans in different order, therefore triggering the Scope close warning messages.

It turns out we have 2 different of grpc servers, the legacy separate and the internal using Vert.x. This can be switched by using this property: `quarkus.grpc.server.use-separate-server=false`

The problem only happens if we use the internal Vert.x based server. 

### Expected behavior

When using the new grpc server we want a unique span for the grpc sever with no warning messages.

### Actual behavior

2 different spans are created and ocasional messages in the logs warning about the Scope not being properly closed.

### How to Reproduce?

Run the GrpcHelloServiceTest on this PR: https://github.com/quarkusio/quarkus/pull/37139 with property `quarkus.grpc.server.use-separate-server=false` at `application.properties`

### Output of `uname -a` or `ver`

Darwin Kernel Version 22.6.0: Thu Nov  2 07:43:57 PDT 2023; root:xnu-8796.141.3.701.17~6/RELEASE_ARM64_T6000

### Output of `java -version`

Java version: 17.0.7, vendor: Eclipse Adoptium

### Quarkus version or git rev

Quarkus 3.5.0 at least

### Build tool (ie. output of `mvnw --version` or `gradlew --version`)

Apache Maven 3.9.5 (57804ffe001d7215b5e7bcb531cf83df38f93546) Maven home: /Users/bbaptist/.sdkman/candidates/maven/current Java version: 17.0.7, vendor: Eclipse Adoptium, Default locale: en_PT, platform encoding: UTF-8 OS name: "mac os x", version: "13.6.2", arch: "aarch64", family: "mac"

**Comment by cescoffier** [comment]
I'm not sure how to handle this as a gRPC message is a HTTP message. Didn't we have a de-duplication feature at some point? 

Alternatively, we could imagine disabling the HTTP instrumentation for gRPC requests. 

@alesj WDYT?

**Comment by alesj** [comment]
> Alternatively, we could imagine disabling the HTTP instrumentation for gRPC requests.

Well, that's the easy way out.
But imho if we do this right, gRPC span should just be a child (or a sub) span from starting HTTP span?
I'll have a closer look asap ... (once I clear my overdue backlog ...)

**Comment by cescoffier** [comment]
Yes, you are right, it should be a child of the HTTP span - at least at the network level.

**Comment by alesj** [comment]
> at least at the network level

Wdym?

**Comment by cescoffier** [comment]
You should get the HTTP span and then a child span for gRPC. 

It's what you have when you use wireshark for example The frame contains HTTP and gRPC-related data.

**Comment by brunobat** [comment]
@cescoffier @alesj like we can see on this draft PR https://github.com/quarkusio/quarkus/pull/37139, we have 2 totally different instrumentations kicking in and they are not related. They cannot remain active at the same time.
If the GRPC span is a child of the HTTP one, to me is secondary and should be fixed later.

**Comment by dan-dr** [comment]
I believe this warning also happens when using `opentelemetry-jdbc` https://quarkus.io/guides/opentelemetry#jdbc

**Comment by dan-dr** [comment]
@brunobat is this likely to happen if I have a JPA operation that happens after the request is returned to the client (i.e. some side effect like incrementing an action counter)?

**Comment by cescoffier** [comment]
@brunobat @alesj what should we do about this one?

**Comment by brunobat** [comment]
@cescoffier We need to address it as part of the Observation API and OTel scope handling fix required for it. We need to have it around.

**Comment by hblab-hann** [comment]
Quarkus 3.15.2 still has this issue

application.properties
```
quarkus.grpc.server.use-separate-server=false
```

log
```
2025-07-08 16:33:36,634 INFO  [io.ope.exp.log.LoggingSpanExporter] (vert.x-eventloop-thread-2) 'POST' : 0ab5d85caa5557395689863609b66aca 6c52ba8f666ed643 SERVER [tracer: io.quarkus.opentelemetry:] AttributesMap{data={http.response.body.size=0, url.scheme=http, url.path=/dummy.v1.DummyService/DummyMethod, http.response.status_code=200, user_agent.original=grpc-node-js/1.11.0-postman.1, client.address=127.0.0.1, http.request.method=POST}, capacity=128, totalAddedValues=7}
2025-07-08 16:33:36,635 INFO  [io.qua.ope.run.QuarkusContextStorage] (vert.x-eventloop-thread-2) Context in storage not the expected context, Scope.close was not called correctly. Details: OTel context before: {spanId=ad696fe739f0b607, traceId=0ab5d85caa5557395689863609b66aca, sampled=true, parentId=6c52ba8f666ed643}. OTel context toAttach: {spanId=6c52ba8f666ed643, traceId=0ab5d85caa5557395689863609b66aca, sampled=true}
2025-07-08 16:33:36,635 INFO  [io.ope.exp.log.LoggingSpanExporter] (executor-thread-1) 'dummy.v1.DummyService/DummyMethod' : 0ab5d85caa5557395689863609b66aca ad696fe739f0b607 SERVER [tracer: io.quarkus.opentelemetry:] AttributesMap{data={rpc.system=grpc, rpc.method=DummyMethod, rpc.grpc.status_code=13, rpc.service=dummy.v1.DummyService}, capacity=128, totalAddedValues=4}
2025-07-08 16:33:36,636 INFO  [io.qua.ope.run.QuarkusContextStorage] (executor-thread-1) Context in storage not the expected context, Scope.close was not called correctly. Details: OTel context before: {}. OTel context toAttach: {spanId=ad696fe739f0b607, traceId=0ab5d85caa5557395689863609b66aca, sampled=true, parentId=6c52ba8f666ed643}
```

**Comment by brunobat** [comment]
Yes, it does.

**Comment by brunobat** [comment]
This user is also affected: https://github.com/quarkusio/quarkus/pull/45478#issuecomment-3046493544
